File: src/C3.Infrastructure/Preferences/LegacyUserSettingsImportResult.vb
Imports System.Collections.Generic

Namespace Preferences

    Public Enum LegacyUserSettingsImportStatus
        NotFound = 0
        Imported
        Failed
    End Enum

    Public NotInheritable Class LegacyUserSettingsImportResult

        Private Sub New(
                status As LegacyUserSettingsImportStatus,
                profile As LegacyUserSettingsProfile,
                rejectedCandidates As IList(Of LegacySettingsProfileReadResult))

            Me.Status = status
            Me.Profile = profile
            Me.RejectedCandidates = New List(Of LegacySettingsProfileReadResult)(
                rejectedCandidates).AsReadOnly()
        End Sub

        Public ReadOnly Property Status As LegacyUserSettingsImportStatus

        Public ReadOnly Property Profile As LegacyUserSettingsProfile

        Public ReadOnly Property RejectedCandidates As IList(Of LegacySettingsProfileReadResult)

        ''' <summary>
        ''' No profile is a successful no-op; only discovered-but-unreadable
        ''' candidates make the import attempt fail.
        ''' </summary>
        Public ReadOnly Property IsSuccess As Boolean
            Get
                Return Status <> LegacyUserSettingsImportStatus.Failed
            End Get
        End Property

        Friend Shared Function NotFound() As LegacyUserSettingsImportResult
            Return New LegacyUserSettingsImportResult(
                LegacyUserSettingsImportStatus.NotFound,
                Nothing,
                New List(Of LegacySettingsProfileReadResult)())
        End Function

        Friend Shared Function Imported(
                profile As LegacyUserSettingsProfile,
                rejectedCandidates As IList(Of LegacySettingsProfileReadResult)) _
                As LegacyUserSettingsImportResult

            Return New LegacyUserSettingsImportResult(
                LegacyUserSettingsImportStatus.Imported,
                profile,
                rejectedCandidates)
        End Function

        Friend Shared Function Failed(
                rejectedCandidates As IList(Of LegacySettingsProfileReadResult)) _
                As LegacyUserSettingsImportResult

            Return New LegacyUserSettingsImportResult(
                LegacyUserSettingsImportStatus.Failed,
                Nothing,
                rejectedCandidates)
        End Function

    End Class

End Namespace

File: src/C3.Infrastructure/Preferences/LegacyUserSettingsImportResult.cs
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace C3.Infrastructure.Preferences
{
	public sealed class LegacyUserSettingsImportResult
	{
		public LegacyUserSettingsImportStatus Status
		{
			get;
		}

		public LegacyUserSettingsProfile Profile
		{
			get;
		}

		public IList<LegacySettingsProfileReadResult> RejectedCandidates
		{
			get;
		}

		public bool IsSuccess => Status != LegacyUserSettingsImportStatus.Failed;

		private LegacyUserSettingsImportResult(LegacyUserSettingsImportStatus status, LegacyUserSettingsProfile profile, IList<LegacySettingsProfileReadResult> rejectedCandidates)
		{
			Status = status;
			Profile = profile;
			RejectedCandidates = new List<LegacySettingsProfileReadResult>(rejectedCandidates).AsReadOnly();
		}

		internal static LegacyUserSettingsImportResult NotFound()
		{
			return new LegacyUserSettingsImportResult(LegacyUserSettingsImportStatus.NotFound, null, new List<LegacySettingsProfileReadResult>());
		}

		internal static LegacyUserSettingsImportResult Imported(LegacyUserSettingsProfile profile, IList<LegacySettingsProfileReadResult> rejectedCandidates)
		{
			return new LegacyUserSettingsImportResult(LegacyUserSettingsImportStatus.Imported, profile, rejectedCandidates);
		}

		internal static LegacyUserSettingsImportResult Failed(IList<LegacySettingsProfileReadResult> rejectedCandidates)
		{
			return new LegacyUserSettingsImportResult(LegacyUserSettingsImportStatus.Failed, null, rejectedCandidates);
		}
	}
}
